voting/src/lib.rs

```diff
@@ -0,0 +1,306 @@
+    // TODO: Revert back to 4 once wasm/wasmer bug is fixed.
+    // See https://github.com/wasmerio/wasmer/issues/1429
+    pub struct U256(8);
+}
+
+/// Voting contract for unlocking transfers. Once the majority of the stake holders agree to
+/// unlock transfer, the time will be recorded and the voting ends.
+#[near_bindgen]
+#[derive(BorshDeserialize, BorshSerialize)]
+pub struct VotingContract {
+    /// How much each validator votes
+    votes: Map<AccountId, Balance>,
+    /// Map of account to their current stake
+    account_stake: Map<AccountId, Balance>,
+    /// Total voted balance so far.
+    total_voted_stake: Balance,
+    /// When the voting ended. `None` means the poll is still open.
+    result: Option<Timestamp>,
+    /// Epoch height when the contract is touched last time.
+    last_epoch_height: EpochHeight,
+}
+
+impl Default for VotingContract {
+    fn default() -> Self {
+        env::panic(b"Voting contract should be initialized before usage")
+    }
+}
+
+#[near_bindgen]
+impl VotingContract {
+    #[init]
+    pub fn new() -> Self {
+        VotingContract {
+            votes: Map::new(b"a".to_vec()),
+            account_stake: Map::new(b"s".to_vec()),
+            total_voted_stake: 0,
+            result: None,
+            last_epoch_height: 0,
+        }
+    }
+
+    fn resolve_votes(&mut self) {
+        let cur_epoch_height = env::epoch_height();
+        if cur_epoch_height != self.last_epoch_height {
+            for account_id in self.account_stake.keys().into_iter().collect::<Vec<_>>() {
+                let old_account_stake = self.account_stake.remove(&account_id).unwrap();
+                let account_current_stake = env::validator_stake(&account_id);
+                let account_voted_stake = self.votes.remove(&account_id).unwrap();
+                if account_current_stake > 0 {
+                    let new_account_voted_stake = (U256::from(account_voted_stake)
+                        * U256::from(account_current_stake)
+                        / U256::from(old_account_stake))
+                    .as_u128();
+                    self.total_voted_stake =
+                        self.total_voted_stake + new_account_voted_stake - account_voted_stake;
+                    self.votes.insert(&account_id, &new_account_voted_stake);
+                    self.account_stake
+                        .insert(&account_id, &account_current_stake);
+                }
+            }
+            self.check_result();
+            self.last_epoch_height = cur_epoch_height;
+        }
+    }
+
+    fn check_result(&mut self) {
+        assert!(
+            self.result.is_none(),
+            "check result is called after result is already set"
+        );
+        let total_stake = env::validator_total_stake();
+        if self.total_voted_stake > 2 * total_stake / 3 {
+            self.result = Some(env::block_timestamp());
+        }
+    }
+
+    /// Vote on a specific proposal with the given stake
```

Comment:
no proposals any more, right? the comment seems wrong.

Also should we even vote with stake amount? If we don't have ability for validators to proxy their delegators votes?

Should it just be that validator can call vote, when they think it's ready to unlock?
And it records and checks their stake at the time of `check_results` call? 
This way I can vote right away, but if people don't think I'm right - they will undelegate to someone else.

This will also be much simpler, because wouldn't need this `resolve_votes` magic.

Reply:
I think this sounds reasonable. @evgenykuzyakov what do you think?